lib/Macros/Sources/SwiftMacros/DebugDescriptionMacro.swift

```diff
@@ -0,0 +1,312 @@
+import SwiftSyntax
+import SwiftSyntaxMacros
+import SwiftDiagnostics
+
+public struct DebugDescriptionMacro {}
+
+/// Attached peer macro which converts `debugDescription` (or `description`) implementations into
+/// an LLDB type summary. The type summary record is emitted into a custom section (via a global
+/// constant), where LLDB will load from at debug time.
+extension DebugDescriptionMacro: PeerMacro {
+    public static func expansion<Decl, Context>(
+        of node: AttributeSyntax,
+        providingPeersOf declaration: Decl,
+        in context: Context
+    )
+    throws -> [DeclSyntax]
+    where Decl: DeclSyntaxProtocol, Context: MacroExpansionContext
+    {
+        // Determine the module name by doing a bit of parsing on fileID.
+        guard let fileID = context.location(of: declaration)?.file,
+              let moduleName = extractModuleName(from: fileID) else {
+            assertionFailure("could not determine module name from fileID")
+            return []
+        }
+
+        guard let typeName = identifier(of: declaration)?.text else {
+            let message: ErrorMessage = "must be attached to a struct/class/enum/extension"
+            context.diagnose(node: node, message: message)
+            return []
+        }
+
+        // Create a lookup for finding properties by name.
+        var properties: [String: PatternBindingSyntax] = [:]
+        for member in memberBlock(declaration)?.members ?? [] {
+            for binding in member.decl.as(VariableDeclSyntax.self)?.bindings ?? [] {
+                if let name = bindingName(binding) {
+                    properties[name] = binding
+                }
+            }
+        }
+
+        // Use `debugDescription` if available, otherwise fallback to `description`.
+        guard let descriptionProperty = properties["debugDescription"] ?? properties["description"] else {
+            let message: ErrorMessage = "debugDescription or description must be defined within \(syntaxKeyword(for: declaration))"
+            context.diagnose(node: declaration, message: message)
+            return []
+        }
+
+        // Validate the body of the description function.
+        //   1. The body must have a single item
+        //   2. The single item must be a string literal
+        //   3. Later on, the interpolation in the string literal will be validated.
+        guard let codeBlock = descriptionProperty.accessorBlock?.accessors.as(CodeBlockItemListSyntax.self),
```

Comment:
Hmm, this will be a bit of an issue. A macro applied to a type shouldn't be able to see the bodies of functions defined within that type, because (for example) we need to be able to determine the members and layout of a type without evaluating the bodies of functions. Unfortunately, @hborla and I discussed this and (IIRC) it's described in a proposal, but the compiler fails to enforce it.

I think what we'll end up needing to do here is collect information about the type itself in a member-attribute macro, then use it to add an attribute to the "description" property which can look at the body. I think that can be made to work.

Reply:
> because (for example) we need to be able to determine the members and layout of a type without evaluating the bodies of functions

does this apply to a peer macro? Since a peer macro doesn't introduce any members, the compiler could determine the members and layout independent of expansion of the peer macro.

Reply:
> I think what we'll end up needing to do here is collect information about the type itself in a member-attribute macro, then use it to add an attribute to the "description" property which can look at the body. I think that can be made to work.

I think this makes sense except for one missing piece to the puzzle. The macro attached do `description` won't have the ability to emit a (global) constant to a custom section. Not currently. Perhaps a change to the compiler could allow a macro attached to `description` to emit data to a custom section. Thoughts?

Reply:
correction: it is now possible to emit static properties to custom sections, thanks to this recent change: https://github.com/apple/swift/pull/69192

Reply:
See e2b3659e186b007df8e28dbb1c6e7ddd5305067c